<html>
<head>
<title>exceptions.py</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">
.s0 { color: #cc7832;}
.s1 { color: #a9b7c6;}
.s2 { color: #6a8759;}
.s3 { color: #629755; font-style: italic;}
.s4 { color: #808080;}
.s5 { color: #6897bb;}
</style>
</head>
<body bgcolor="#2b2b2b">
<table CELLSPACING=0 CELLPADDING=5 COLS=1 WIDTH="100%" BGCOLOR="#606060" >
<tr><td><center>
<font face="Arial, Helvetica" color="#000000">
exceptions.py</font>
</center></td></tr></table>
<pre><span class="s0">class </span><span class="s1">PlotlyError(Exception):</span>
    <span class="s0">pass</span>


<span class="s0">class </span><span class="s1">PlotlyEmptyDataError(PlotlyError):</span>
    <span class="s0">pass</span>


<span class="s0">class </span><span class="s1">PlotlyGraphObjectError(PlotlyError):</span>
    <span class="s0">def </span><span class="s1">__init__(self</span><span class="s0">, </span><span class="s1">message=</span><span class="s2">&quot;&quot;</span><span class="s0">, </span><span class="s1">path=()</span><span class="s0">, </span><span class="s1">notes=()):</span>
        <span class="s3">&quot;&quot;&quot; 
        General graph object error for validation failures. 
 
        :param (str|unicode) message: The error message. 
        :param (iterable) path: A path pointing to the error. 
        :param notes: Add additional notes, but keep default exception message. 
 
        &quot;&quot;&quot;</span>
        <span class="s1">self.message = message</span>
        <span class="s1">self.plain_message = message  </span><span class="s4"># for backwards compat</span>
        <span class="s1">self.path = list(path)</span>
        <span class="s1">self.notes = notes</span>
        <span class="s1">super(PlotlyGraphObjectError</span><span class="s0">, </span><span class="s1">self).__init__(message)</span>

    <span class="s0">def </span><span class="s1">__str__(self):</span>
        <span class="s3">&quot;&quot;&quot;This is called by Python to present the error message.&quot;&quot;&quot;</span>
        <span class="s1">format_dict = {</span>
            <span class="s2">&quot;message&quot;</span><span class="s1">: self.message</span><span class="s0">,</span>
            <span class="s2">&quot;path&quot;</span><span class="s1">: </span><span class="s2">&quot;[&quot; </span><span class="s1">+ </span><span class="s2">&quot;][&quot;</span><span class="s1">.join(repr(k) </span><span class="s0">for </span><span class="s1">k </span><span class="s0">in </span><span class="s1">self.path) + </span><span class="s2">&quot;]&quot;</span><span class="s0">,</span>
            <span class="s2">&quot;notes&quot;</span><span class="s1">: </span><span class="s2">&quot;</span><span class="s0">\n</span><span class="s2">&quot;</span><span class="s1">.join(self.notes)</span><span class="s0">,</span>
        <span class="s1">}</span>
        <span class="s0">return </span><span class="s2">&quot;{message}</span><span class="s0">\n\n</span><span class="s2">Path To Error: {path}</span><span class="s0">\n\n</span><span class="s2">{notes}&quot;</span><span class="s1">.format(**format_dict)</span>


<span class="s0">class </span><span class="s1">PlotlyDictKeyError(PlotlyGraphObjectError):</span>
    <span class="s0">def </span><span class="s1">__init__(self</span><span class="s0">, </span><span class="s1">obj</span><span class="s0">, </span><span class="s1">path</span><span class="s0">, </span><span class="s1">notes=()):</span>
        <span class="s3">&quot;&quot;&quot;See PlotlyGraphObjectError.__init__ for param docs.&quot;&quot;&quot;</span>
        <span class="s1">format_dict = {</span><span class="s2">&quot;attribute&quot;</span><span class="s1">: path[-</span><span class="s5">1</span><span class="s1">]</span><span class="s0">, </span><span class="s2">&quot;object_name&quot;</span><span class="s1">: obj._name}</span>
        <span class="s1">message = </span><span class="s2">&quot;'{attribute}' is not allowed in '{object_name}'&quot;</span><span class="s1">.format(</span>
            <span class="s1">**format_dict</span>
        <span class="s1">)</span>
        <span class="s1">notes = [obj.help(return_help=</span><span class="s0">True</span><span class="s1">)] + list(notes)</span>
        <span class="s1">super(PlotlyDictKeyError</span><span class="s0">, </span><span class="s1">self).__init__(</span>
            <span class="s1">message=message</span><span class="s0">, </span><span class="s1">path=path</span><span class="s0">, </span><span class="s1">notes=notes</span>
        <span class="s1">)</span>


<span class="s0">class </span><span class="s1">PlotlyDictValueError(PlotlyGraphObjectError):</span>
    <span class="s0">def </span><span class="s1">__init__(self</span><span class="s0">, </span><span class="s1">obj</span><span class="s0">, </span><span class="s1">path</span><span class="s0">, </span><span class="s1">notes=()):</span>
        <span class="s3">&quot;&quot;&quot;See PlotlyGraphObjectError.__init__ for param docs.&quot;&quot;&quot;</span>
        <span class="s1">format_dict = {</span><span class="s2">&quot;attribute&quot;</span><span class="s1">: path[-</span><span class="s5">1</span><span class="s1">]</span><span class="s0">, </span><span class="s2">&quot;object_name&quot;</span><span class="s1">: obj._name}</span>
        <span class="s1">message = </span><span class="s2">&quot;'{attribute}' has invalid value inside '{object_name}'&quot;</span><span class="s1">.format(</span>
            <span class="s1">**format_dict</span>
        <span class="s1">)</span>
        <span class="s1">notes = [obj.help(path[-</span><span class="s5">1</span><span class="s1">]</span><span class="s0">, </span><span class="s1">return_help=</span><span class="s0">True</span><span class="s1">)] + list(notes)</span>
        <span class="s1">super(PlotlyDictValueError</span><span class="s0">, </span><span class="s1">self).__init__(</span>
            <span class="s1">message=message</span><span class="s0">, </span><span class="s1">notes=notes</span><span class="s0">, </span><span class="s1">path=path</span>
        <span class="s1">)</span>


<span class="s0">class </span><span class="s1">PlotlyListEntryError(PlotlyGraphObjectError):</span>
    <span class="s0">def </span><span class="s1">__init__(self</span><span class="s0">, </span><span class="s1">obj</span><span class="s0">, </span><span class="s1">path</span><span class="s0">, </span><span class="s1">notes=()):</span>
        <span class="s3">&quot;&quot;&quot;See PlotlyGraphObjectError.__init__ for param docs.&quot;&quot;&quot;</span>
        <span class="s1">format_dict = {</span><span class="s2">&quot;index&quot;</span><span class="s1">: path[-</span><span class="s5">1</span><span class="s1">]</span><span class="s0">, </span><span class="s2">&quot;object_name&quot;</span><span class="s1">: obj._name}</span>
        <span class="s1">message = </span><span class="s2">&quot;Invalid entry found in '{object_name}' at index, '{index}'&quot;</span><span class="s1">.format(</span>
            <span class="s1">**format_dict</span>
        <span class="s1">)</span>
        <span class="s1">notes = [obj.help(return_help=</span><span class="s0">True</span><span class="s1">)] + list(notes)</span>
        <span class="s1">super(PlotlyListEntryError</span><span class="s0">, </span><span class="s1">self).__init__(</span>
            <span class="s1">message=message</span><span class="s0">, </span><span class="s1">path=path</span><span class="s0">, </span><span class="s1">notes=notes</span>
        <span class="s1">)</span>


<span class="s0">class </span><span class="s1">PlotlyDataTypeError(PlotlyGraphObjectError):</span>
    <span class="s0">def </span><span class="s1">__init__(self</span><span class="s0">, </span><span class="s1">obj</span><span class="s0">, </span><span class="s1">path</span><span class="s0">, </span><span class="s1">notes=()):</span>
        <span class="s3">&quot;&quot;&quot;See PlotlyGraphObjectError.__init__ for param docs.&quot;&quot;&quot;</span>
        <span class="s1">format_dict = {</span><span class="s2">&quot;index&quot;</span><span class="s1">: path[-</span><span class="s5">1</span><span class="s1">]</span><span class="s0">, </span><span class="s2">&quot;object_name&quot;</span><span class="s1">: obj._name}</span>
        <span class="s1">message = </span><span class="s2">&quot;Invalid entry found in '{object_name}' at index, '{index}'&quot;</span><span class="s1">.format(</span>
            <span class="s1">**format_dict</span>
        <span class="s1">)</span>
        <span class="s1">note = </span><span class="s2">&quot;It's invalid because it doesn't contain a valid 'type' value.&quot;</span>
        <span class="s1">notes = [note] + list(notes)</span>
        <span class="s1">super(PlotlyDataTypeError</span><span class="s0">, </span><span class="s1">self).__init__(</span>
            <span class="s1">message=message</span><span class="s0">, </span><span class="s1">path=path</span><span class="s0">, </span><span class="s1">notes=notes</span>
        <span class="s1">)</span>


<span class="s0">class </span><span class="s1">PlotlyKeyError(KeyError):</span>
    <span class="s3">&quot;&quot;&quot; 
    KeyErrors are not printed as beautifully as other errors (this is so that 
    {}[''] prints    &quot;KeyError: ''&quot; and not &quot;KeyError:&quot;). So here we use 
    LookupError's __str__ to make a PlotlyKeyError object which will print nicer 
    error messages for KeyErrors. 
    &quot;&quot;&quot;</span>

    <span class="s0">def </span><span class="s1">__str__(self):</span>
        <span class="s0">return </span><span class="s1">LookupError.__str__(self)</span>
</pre>
</body>
</html>